components/experimental/src/duration/format.rs

```diff
@@ -377,17 +420,209 @@ impl<'a> FormattedDuration<'a> {
+        // a. Set listStyle to "short".
+
+        // 5. Perform ! CreateDataPropertyOrThrow(lfOpts, "style", listStyle).
+        // 6. Let lf be ! Construct(%ListFormat%, « durationFormat.[[Locale]], lfOpts »).
+
+        // Note: the above steps are performed while initializing DurationFormatter.
+        let formatted_list = self.fmt.list.format(parts_list.iter().map(|s| &s.string));
+
+        let mut list_sink = PartSink::new();
+        formatted_list.write_to_parts(&mut list_sink)?;
+
+        // 7. Let strings be a new empty List.
+        // 8. For each element parts of partitionedPartsList, do
+
+        // a. Let string be the empty String.
+        // b. For each Record { [[Type]], [[Value]], [[Unit]] } part in parts, do
+        //     i. Set string to the string-concatenation of string and part.[[Value]].
+        // c. Append string to strings.
+
+        // 9. Let formattedPartsList be CreatePartsFromList(lf, strings).
+        // 10. Let partitionedPartsIndex be 0.
+        let mut partitioned_parts_index = 0;
+        // 11. Let partitionedLength be the number of elements in partitionedPartsList.
+        let partitioned_length = parts_list.len();
+        // 12. Let flattenedPartsList be a new empty List.
+        // 13. For each Record { [[Type]], [[Value]] } listPart in formattedPartsList, do
+
+        let PartSink {
+            string: list_str,
+            parts: list_parts,
+        } = list_sink;
+
+        for (start, end, list_part) in list_parts {
+            // a. If listPart.[[Type]] is "element", then
+            if list_part == icu_list::parts::ELEMENT {
+                // i. Assert: partitionedPartsIndex < partitionedLength.
+                debug_assert!(partitioned_parts_index < partitioned_length);
+                // ii. Let parts be partitionedPartsList[partitionedPartsIndex].
+                let PartSink { string, parts } = &parts_list[partitioned_parts_index];
+                // iii. For each Record { [[Type]], [[Value]], [[Unit]] } part in parts, do
+                for (start, end, part) in parts {
+                    // 1. Append part to flattenedPartsList.
+                    sink.with_part(*part, |w| w.write_str(&string[*start..*end]))?;
+                }
+                // iv. Set partitionedPartsIndex to partitionedPartsIndex + 1.
+                partitioned_parts_index += 1;
+            }
+            // b. Else,
+            else {
+                // i. Assert: listPart.[[Type]] is "literal".
+                debug_assert!(list_part == icu_list::parts::LITERAL);
+                // ii. Append the Record { [[Type]]: "literal", [[Value]]: listPart.[[Value]], [[Unit]]: empty } to flattenedPartsList.
+                sink.with_part(parts::LITERAL, |w| w.write_str(&list_str[start..end]))?;
+            }
+        }
+
+        // 14. Return flattenedPartsList.
+        Ok(())
+    }
+}
+
+struct PartSink {
+    string: String,
+    parts: Vec<(usize, usize, writeable::Part)>,
+}
+
```

Comment:
`PartSink` is very similar to the `TestSink` the `writeable` library internally uses. I create it here to add part information to ListFormatter-emitted parts. (changing `parts::Element` to `parts::HOUR`, etc.)

It would be nice if ListFormatter could optionally preserve part annotations for the Writeables it formats. I'd be happy to try to work on that if that sounds useful.

Reply:
Good point, @sffc , what do you think ?

Reply:
You shouldn't ever need to allocate anything, including into a `Vec` or `String`, when formatting in ICU4X. If we can do it in datetime format, which is an order of magnitude more complex than any other ICU4X formatter right now, we can do it here in DurationFormatter. :)

Please note that ListFormatter is already wired up to allow you to do this. The ListFormatter invokes the `write_to_parts` function of the writeables passed to its `format` function. You should make sure those writeables add the `unit` part. You'll get the `ELEMENT` and `FIELD` parts from `icu_list`, but that's fine.